x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/bucket/histogram/HistoBackedHistogramAggregator.java

```diff
@@ -0,0 +1,93 @@
+import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.search.aggregations.CardinalityUpperBound;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.bucket.histogram.AbstractHistogramAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.xpack.analytics.aggregations.support.HistogramValuesSource;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class HistoBackedHistogramAggregator extends AbstractHistogramAggregator {
+
+    private final HistogramValuesSource.Histogram valuesSource;
+
+    public HistoBackedHistogramAggregator(
+        String name,
+        AggregatorFactories factories,
+        double interval,
+        double offset,
+        BucketOrder order,
+        boolean keyed,
+        long minDocCount,
+        double minBound,
+        double maxBound,
+        ValuesSourceConfig valuesSourceConfig,
+        SearchContext context,
+        Aggregator parent,
+        CardinalityUpperBound cardinalityUpperBound,
+        Map<String, Object> metadata) throws IOException {
+        super(name, factories, interval, offset, order, keyed, minDocCount, minBound, maxBound,
+            valuesSourceConfig.format(), context, parent, cardinalityUpperBound, metadata);
+
+        // TODO: Stop using null here
+        this.valuesSource = valuesSourceConfig.hasValues() ? (HistogramValuesSource.Histogram) valuesSourceConfig.getValuesSource() : null;
+
+        // Sub aggregations are not allowed when running histogram agg over histograms
+        if (subAggregators().length > 0) {
+            throw new IllegalArgumentException("Histogram aggregation on histogram fields does not support sub-aggregations");
+        }
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+
+        final HistogramValues values = valuesSource.getHistogramValues(ctx);
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                if (values.advanceExact(doc)) {
+                    final HistogramValue sketch = values.histogram();
+
+                    double previousKey = Double.NEGATIVE_INFINITY;
+                    while (sketch.next()) {
+                        final double value = sketch.value();
+                        final int count = sketch.count();
+
+                        double key = Math.floor((value - offset) / interval);
+                        assert key >= previousKey;
+                        for (int i = 0; i < count; ++i) {
```

Comment:
I think rather than iterating `count` times, we can lookup the bucketOrd once, call `collectExistingBucket()` / `collectBucket()` as appropriate, and then finally call `incrementBucketDocCount()` with count-1.  That way we don't have to keep re-adding to the same ord just to increment the count, and should be a bit faster I think.